src/wp-includes/class-wp-block.php

```diff
@@ -506,14 +532,19 @@ public function render( $options = array() ) {
 					if ( ! is_null( $pre_render ) ) {
 						$block_content .= $pre_render;
 					} else {
-						$source_block = $inner_block->parsed_block;
+						$source_block        = $inner_block->parsed_block;
+						$inner_block_context = $inner_block->context;
 
 						/** This filter is documented in wp-includes/blocks.php */
 						$inner_block->parsed_block = apply_filters( 'render_block_data', $inner_block->parsed_block, $source_block, $parent_block );
 
 						/** This filter is documented in wp-includes/blocks.php */
 						$inner_block->context = apply_filters( 'render_block_context', $inner_block->context, $inner_block->parsed_block, $parent_block );
 
+						if ( $inner_block->context !== $inner_block_context ) {
```

Comment:
Related to https://github.com/WordPress/wordpress-develop/pull/7522#discussion_r1792139774, we also need to consider `$inner_block->parsed_block` here.
```suggestion
						if ( $inner_block->parsed_block !== $source_block || $inner_block->context !== $inner_block_context ) {
```